### DISCUSSION TITLE
Parallelisation of NLP with ML surrogates embedded

### DISCUSSION THREAD

**Discussion by mawbray** [casadi/casadi#3683] [Q&A] [answered]
Hi casadi community! 

I am working on an alg. that requires solving many NLP with ML surrogates embedded in reduced form as equality constraints. 
I have had some success embedding constraints and passing to IPOPT via casadi callbacks making use of tensorflow AD. 
Ideally I would parallelise but realise tensorflow does not play too nicely with multiprocessing. 
Could you suggest a best approach here?

I have had a search for this elsewhere but not sure anything has been asked that covers these bases.
Thanks for your help in advance,
Max

**Comment by ernestds** [comment]
I have nothing useful to say, but this might be of use? https://github.com/Tim-Salzmann/l4casadi

**Reply by mawbray** [reply]
Thanks @ernestds ! I was aware of L4Casadi.  For various reasons I committed to using JAX fairly early on, and it is straightforward to wrap jax models to tensorflow, e.g. https://www.tensorflow.org/guide/jax2tf. I think for that reason I am holding off on L4Casadi and pytorch unless there are no other options.

**Reply by ernestds** [reply, marked as answer]
Maybe something like Ray? https://docs.ray.io/en/latest/index.html
So if I understand it correctly, you currently solve one NLP problem which has a Tensorflow model in a Casadi callback as a constraint right? Then you want to solve multiple of those NLPs in parallel.

Do you think something like this could work?
https://stackoverflow.com/questions/36610290/tensorflow-and-multiprocessing-passing-sessions#54562773
For example if you make the NLP problem in the Simulator class, in the example, or something like this.

I have no experience with this but it does feel like something I'll eventually come across. Let me know if you find a solution :)

**Reply by mawbray** [reply]
Thanks Ernesto, this looks like a good option - thanks for the suggestion! I will let you know how it goes :)

**Reply by mawbray** [reply]
Just an update on this - ray was the way to go. Thanks @ernestds for the help :)
